File: openmrs-client/src/main/java/org/openmrs/mobile/api/EncounterService.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */

package org.openmrs.mobile.api;

import android.app.IntentService;
import android.content.Intent;

import org.openmrs.mobile.api.repository.VisitRepository;
import org.openmrs.mobile.dao.PatientDAO;
import org.openmrs.mobile.dao.VisitDAO;
import org.openmrs.mobile.listeners.retrofit.DefaultResponseCallbackListener;
import org.openmrs.mobile.listeners.retrofit.StartVisitResponseListenerCallback;
import org.openmrs.mobile.models.Encounter;
import org.openmrs.mobile.models.EncounterType;
import org.openmrs.mobile.models.Encountercreate;
import org.openmrs.mobile.utilities.ActiveAndroid.query.Select;
import org.openmrs.mobile.utilities.NetworkUtils;
import org.openmrs.mobile.utilities.ToastUtil;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import rx.android.schedulers.AndroidSchedulers;

public class EncounterService extends IntentService {

    private final RestApi apiService = RestServiceBuilder.createService(RestApi.class);

    public EncounterService() {
        super("Save Encounter");
    }

    public void addEncounter(final Encountercreate encountercreate, @Nullable DefaultResponseCallbackListener callbackListener) {

        if(NetworkUtils.isOnline()) {
            new VisitDAO().getActiveVisitByPatientId(encountercreate.getPatientId())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(visit -> {
                        if (visit != null) {
                            encountercreate.setVisit(visit.getUuid());
                            if (callbackListener != null) {
                                syncEncounter(encountercreate, callbackListener);
                            }
                            else {
                                syncEncounter(encountercreate);
                            }
                        } else {

                            startNewVisitForEncounter(encountercreate);
                        }
                    });
        }
        else
            ToastUtil.error("No internet connection. Form data is saved locally " +
                    "and will sync when internet connection is restored. ");
    }

    public void addEncounter(final Encountercreate encountercreate) {
        addEncounter(encountercreate, null);
    }

        private void startNewVisitForEncounter(final Encountercreate encountercreate, @Nullable final DefaultResponseCallbackListener callbackListener) {
        new VisitRepository().startVisit(new PatientDAO().findPatientByUUID(encountercreate.getPatient()),
                new StartVisitResponseListenerCallback() {
                    @Override
                    public void onStartVisitResponse(long id) {
                        new VisitDAO().getVisitByID(id)
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe(visit -> {
                                    encountercreate.setVisit(visit.getUuid());
                                    if (callbackListener != null) {
                                        syncEncounter(encountercreate, callbackListener);
                                    }
                                    else {
                                        syncEncounter(encountercreate);
                                    }
                                });
                    }
                    @Override
                    public void onResponse() {
                        // This method is intentionally empty
                    }
                    @Override
                    public void onErrorResponse(String errorMessage) {
                        ToastUtil.error(errorMessage);
                    }
                });
    }

    public void startNewVisitForEncounter(final Encountercreate encountercreate) {
        startNewVisitForEncounter(encountercreate, null);
    }

    public void syncEncounter(final Encountercreate encountercreate, @Nullable final DefaultResponseCallbackListener callbackListener) {

        if (NetworkUtils.isOnline()) {

            encountercreate.pullObslist();
            Call<Encounter> call = apiService.createEncounter(encountercreate);
            call.enqueue(new Callback<Encounter>() {
                @Override
                public void onResponse(@NonNull Call<Encounter> call, @NonNull Response<Encounter> response) {
                    if (response.isSuccessful()) {
                        Encounter encounter = response.body();
                        linkvisit(encountercreate.getPatientId(),encountercreate.getFormname(), encounter, encountercreate);
                        encountercreate.setSynced(true);
                        encountercreate.save();
                        new VisitRepository().syncLastVitals(encountercreate.getPatient());
                        if (callbackListener != null) {
                            callbackListener.onResponse();
                        }
                    } else {
                        if (callbackListener != null) {
                            callbackListener.onErrorResponse(response.errorBody().toString());
                        }
                    }
                }

                @Override
                public void onFailure(@NonNull Call<Encounter> call, @NonNull Throwable t) {
                    if (callbackListener != null) {
                        callbackListener.onErrorResponse(t.getLocalizedMessage());
                    }
                }
            });

        } else {
            ToastUtil.error("Sync is off. Turn on sync to save form data.");
        }

    }

    public void syncEncounter(final Encountercreate encountercreate) {
        syncEncounter(encountercreate, null);
    }

    private void linkvisit(Long patientid, String formname, Encounter encounter, Encountercreate encountercreate)
    {
        VisitDAO visitDAO = new VisitDAO();
        visitDAO.getVisitByUuid(encounter.getVisit().getUuid())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(visit -> {
                    encounter.setEncounterType(new EncounterType(formname));
                    for (int i=0;i<encountercreate.getObservations().size();i++)
                    {
                        encounter.getObservations().get(i).setDisplayValue
                                (encountercreate.getObservations().get(i).getValue());
                    }
                    List<Encounter> encounterList=visit.getEncounters();
                    encounterList.add(encounter);
                    visitDAO.saveOrUpdate(visit, patientid)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(id ->
                                    ToastUtil.success(formname+" data saved successfully"));
                });
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        if(NetworkUtils.isOnline()) {

            List<Encountercreate> encountercreatelist = new Select()
                    .from(Encountercreate.class)
                    .execute();

            for(final Encountercreate encountercreate:encountercreatelist)
            {
                if(!encountercreate.getSynced() &&
                        new PatientDAO().findPatientByID(Long.toString(encountercreate.getPatientId())).isSynced())
                {
                    new VisitDAO().getActiveVisitByPatientId(encountercreate.getPatientId())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(visit -> {
                                if (visit != null) {
                                    encountercreate.setVisit(visit.getUuid());
                                    syncEncounter(encountercreate);

                                } else {
                                    startNewVisitForEncounter(encountercreate);
                                }
                            });
                }
            }


        } else {
            ToastUtil.error("No internet connection. Form data is saved locally " +
                    "and will sync when internet connection is restored. ");
        }
    }

}

File: openmrs-client/src/main/java/org/openmrs/mobile/api/EncounterService.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs.mobile.api

import android.app.IntentService
import android.content.Intent
import org.openmrs.mobile.R
import org.openmrs.mobile.api.repository.VisitRepository
import org.openmrs.mobile.application.OpenMRS
import org.openmrs.mobile.dao.PatientDAO
import org.openmrs.mobile.dao.VisitDAO
import org.openmrs.mobile.databases.AppDatabase
import org.openmrs.mobile.listeners.retrofitcallbacks.DefaultResponseCallback
import org.openmrs.mobile.listeners.retrofitcallbacks.StartVisitResponseCallback
import org.openmrs.mobile.models.Encounter
import org.openmrs.mobile.models.EncounterType
import org.openmrs.mobile.models.Encountercreate
import org.openmrs.mobile.models.Visit
import org.openmrs.mobile.utilities.NetworkUtils.isOnline
import org.openmrs.mobile.utilities.ToastUtil.error
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import rx.android.schedulers.AndroidSchedulers

class EncounterService : IntentService("Save Encounter") {
    private val apiService = RestServiceBuilder.createService(RestApi::class.java)
    @JvmOverloads
    fun addEncounter(encountercreate: Encountercreate, callbackListener: DefaultResponseCallback? = null) {
        if (isOnline()) {
            VisitDAO().getActiveVisitByPatientId(encountercreate.patientId)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe { visit: Visit? ->
                        if (visit != null) {
                            encountercreate.visit = visit.uuid
                            if (callbackListener != null) {
                                syncEncounter(encountercreate, callbackListener)
                            } else {
                                syncEncounter(encountercreate)
                            }
                        } else {
                            startNewVisitForEncounter(encountercreate)
                        }
                    }
        } else {
            error(getString(R.string.form_data_will_be_synced_later_error_message))
        }
    }

    private fun startNewVisitForEncounter(encountercreate: Encountercreate, callbackListener: DefaultResponseCallback?) {
        VisitRepository().startVisit(PatientDAO().findPatientByUUID(encountercreate.patient),
                object : StartVisitResponseCallback {
                    override fun onStartVisitResponse(id: Long) {
                        VisitDAO().getVisitByID(id)
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribe { visit: Visit ->
                                    encountercreate.visit = visit.uuid
                                    if (callbackListener != null) {
                                        syncEncounter(encountercreate, callbackListener)
                                    } else {
                                        syncEncounter(encountercreate)
                                    }
                                }
                    }

                    override fun onResponse() {
                        // This method is intentionally empty
                    }

                    override fun onErrorResponse(errorMessage: String) {
                        error(errorMessage)
                    }
                })
    }

    fun startNewVisitForEncounter(encountercreate: Encountercreate) {
        startNewVisitForEncounter(encountercreate, null)
    }

    @JvmOverloads
    fun syncEncounter(encountercreate: Encountercreate, callbackListener: DefaultResponseCallback? = null) {
        if (isOnline()) {
            val call = apiService.createEncounter(encountercreate)
            call.enqueue(object : Callback<Encounter?> {
                override fun onResponse(call: Call<Encounter?>, response: Response<Encounter?>) {
                    if (response.isSuccessful) {
                        val encounter = response.body()
                        linkvisit(encountercreate.patientId, encountercreate.formname, encounter, encountercreate)
                        encountercreate.synced = true
                        AppDatabase.getDatabase(OpenMRS.getInstance().applicationContext)
                                .encounterCreateRoomDAO()
                                .updateExistingEncounter(encountercreate)
                        VisitRepository().syncLastVitals(encountercreate.patient)
                        callbackListener?.onResponse()
                    } else {
                        callbackListener?.onErrorResponse(response.errorBody().toString())
                    }
                }

                override fun onFailure(call: Call<Encounter?>, t: Throwable) {
                    callbackListener?.onErrorResponse(t.localizedMessage)
                }
            })
        } else {
            error(getString(R.string.form_data_sync_is_off_message))
        }
    }

    private fun linkvisit(patientid: Long?, formname: String?, encounter: Encounter?, encountercreate: Encountercreate) {
        val visitDAO = VisitDAO()
        visitDAO.getVisitByUuid(encounter!!.visit!!.uuid)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { visit: Visit ->
                    encounter.encounterType = EncounterType(formname)
                    for (i in encountercreate.observations.indices) {
                        encounter.observations[i].displayValue = encountercreate.observations[i].value
                    }
                    val encounterList: MutableList<Encounter?> = visit.encounters as MutableList<Encounter?>
                    encounterList.add(encounter)
                    visitDAO.saveOrUpdate(visit, patientid!!)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe()
                }
    }

    override fun onHandleIntent(intent: Intent?) {
        if (isOnline()) {
            val encountercreatelist = AppDatabase.getDatabase(OpenMRS.getInstance().applicationContext)
                    .encounterCreateRoomDAO()
                    .allCreatedEncounters
            for (encountercreate in encountercreatelist) {
                if (!encountercreate.synced &&
                        PatientDAO().findPatientByID(java.lang.Long.toString(encountercreate.patientId!!)).isSynced) {
                    VisitDAO().getActiveVisitByPatientId(encountercreate.patientId)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe { visit: Visit? ->
                                if (visit != null) {
                                    encountercreate.visit = visit.uuid
                                    syncEncounter(encountercreate)
                                } else {
                                    startNewVisitForEncounter(encountercreate)
                                }
                            }
                }
            }
        } else {
            error(getString(R.string.form_data_will_be_synced_later_error_message))
        }
    }
}